gcs/gcs.go

```diff
@@ -75,18 +92,46 @@ func newFilter(version uint16, B uint8, M uint64, key [KeySize]byte, data [][]by
 		panic(fmt.Sprintf("B value of %d is greater than max allowed 32", B))
 	}
 	switch version {
-	case 1:
+	case 1, 2:
 	default:
 		panic(fmt.Sprintf("version %d filters are not supported", version))
 	}
 
+	// Note that some of the entries might end up hashing to duplicates and
+	// being removed below, so it is important to perform this check on the
+	// raw data to maintain consensus.
 	numEntries := uint64(len(data))
 	if numEntries > math.MaxInt32 {
 		str := fmt.Sprintf("unable to create filter with %d entries greater "+
 			"than max allowed %d", len(data), math.MaxInt32)
 		return nil, makeError(ErrNTooBig, str)
 	}
 
+	// Insert the hash of each data element into a slice while removing any
+	// duplicates in the data for filters after version 1.
+	k0 := binary.LittleEndian.Uint64(key[0:8])
+	k1 := binary.LittleEndian.Uint64(key[8:16])
+	values := make([]uint64, 0, numEntries)
+	switch {
+	case version == 1:
+		for _, d := range data {
+			v := siphash.Hash(k0, k1, d)
+			values = append(values, v)
+		}
+
+	case version > 1:
+		seen := make(map[uint64]struct{}, numEntries*2)
```

Comment:
This was probably benchmarked to come to this decision, but why multiply numEntries by 2 for the initial capacity? The way the initial capacity works on maps already takes into account the increased number of hashmap buckets required to keep the amortized O(n) lookups.

Reply:
Correct, it is the result of benchmarking a bunch of actual blocks on the network.  I found the result quite strange as well as my understanding of maps is precisely what you describe, but benchmarking shows there is a massive difference (3x) in number of allocations when only specifying the actual size.